File: src/claims/claims.ts
/**
 * @nymrel/swarm-protocol - Resource Claims & Auto-Expiring Leases
 * Copyright (c) 2026 Nymrel / JalenBuilds LLC. Licensed under the MIT License.
 */

import * as fs from 'node:fs';
import * as path from 'node:path';
import * as crypto from 'node:crypto';
import {
  ClaimRecord,
  AcquireClaimParams,
  ClaimFilter,
  ClaimMode,
  ClaimStatus,
} from '../types';
import { AtomicLockManager } from '../fencing/lock';
import { FencingClock } from '../fencing/fencing';
import { FileMailboxManager } from '../bus/mailbox';

export class ClaimManager {
  private baseDir: string;
  private lockManager: AtomicLockManager;
  private fencingClock: FencingClock;
  private mailboxManager: FileMailboxManager;
  private defaultLeaseMs: number;

  constructor(swarmRoot: string, defaultLeaseMs = 30000) {
    this.baseDir = path.join(swarmRoot, 'claims');
    this.lockManager = new AtomicLockManager(swarmRoot);
    this.fencingClock = new FencingClock(swarmRoot);
    this.mailboxManager = new FileMailboxManager(swarmRoot);
    this.defaultLeaseMs = defaultLeaseMs;

    this.ensureDir(this.baseDir);
  }

  private ensureDir(dirPath: string): void {
    if (!fs.existsSync(dirPath)) {
      fs.mkdirSync(dirPath, { recursive: true });
    }
  }

  private getClaimPath(claimId: string): string {
    return path.join(this.baseDir, `${claimId}.json`);
  }

  /**
   * Normalize path for cross-platform hierarchy checks.
   */
  static normalizeResourcePath(resourcePath: string): string {
    let normalized = resourcePath.replace(/\\/g, '/').trim();
    if (normalized.endsWith('/') && normalized.length > 1) {
      normalized = normalized.slice(0, -1);
    }
    // Lowercase for Windows path case-insensitivity
    return normalized.toLowerCase();
  }

  /**
   * Check if path A is an ancestor or exact match of path B.
   */
  static isSubPath(parent: string, child: string): boolean {
    if (parent === child) return true;
    const parentWithSlash = parent.endsWith('/') ? parent : `${parent}/`;
    return child.startsWith(parentWithSlash);
  }

  /**
   * Check if two claims conflict based on path hierarchy and mode.
   */
  static checkConflict(
    pathA: string,
    modeA: ClaimMode,
    pathB: string,
    modeB: ClaimMode
  ): boolean {
    const isRelated = this.isSubPath(pathA, pathB) || this.isSubPath(pathB, pathA);
    if (!isRelated) {
      return false;
    }

    // If either claim is exclusive, any hierarchy overlap is a conflict
    if (modeA === 'exclusive' || modeB === 'exclusive') {
      return true;
    }

    // Both shared claims on related paths can co-exist
    return false;
  }

  /**
   * Acquire a new resource claim with lease and fencing generation.
   */
  async acquireClaim(params: AcquireClaimParams): Promise<ClaimRecord> {
    return await this.lockManager.withLock('claims_mutex', async () => {
      const normPath = ClaimManager.normalizeResourcePath(params.resource_path);
      const mode = params.mode ?? 'exclusive';
      const leaseDurationMs = params.lease_duration_ms ?? this.defaultLeaseMs;
      const nowMs = Date.now();
      const nowIso = new Date(nowMs).toISOString();
      const expiresIso = new Date(nowMs + leaseDurationMs).toISOString();

      // Read active claims and check for conflicts
      const activeClaims = await this.listActiveClaimsInternal();

      for (const existing of activeClaims) {
        // Skip same agent re-claiming exact same path
        if (existing.owner_agent === params.owner_agent && existing.resource_path === normPath) {
          // Refresh existing claim
          existing.lease_duration_ms = leaseDurationMs;
          existing.expires_at = expiresIso;
          existing.heartbeat_at = nowIso;
          existing.metadata = { ...existing.metadata, ...params.metadata };
          fs.writeFileSync(this.getClaimPath(existing.claim_id), JSON.stringify(existing, null, 2), 'utf-8');
          return existing;
        }

        const conflict = ClaimManager.checkConflict(
          existing.resource_path,
          existing.mode,
          normPath,
          mode
        );

        if (conflict) {
          throw new Error(
            `Claim conflict on "${params.resource_path}": Already held by "${existing.owner_agent}" ` +
            `(${existing.mode} mode, claim_id: ${existing.claim_id}, expires: ${existing.expires_at})`
          );
        }
      }

      // Generate monotonic fencing token for the resource
      const fencingToken = await this.fencingClock.incrementGeneration(normPath, params.owner_agent);

      const claimId = crypto.randomUUID();
      const record: ClaimRecord = {
        claim_id: claimId,
        resource_path: normPath,
        owner_agent: params.owner_agent,
        mode,
        status: 'active',
        fencing_generation: fencingToken.generation,
        lease_duration_ms: leaseDurationMs,
        acquired_at: nowIso,
        expires_at: expiresIso,
        heartbeat_at: nowIso,
        metadata: params.metadata,
      };

      fs.writeFileSync(this.getClaimPath(claimId), JSON.stringify(record, null, 2), 'utf-8');

      // Record event
      await this.mailboxManager.recordEvent({
        event_id: crypto.randomUUID(),
        timestamp: nowIso,
        event_type: 'claim_acquired',
        actor: params.owner_agent,
        resource: normPath,
        details: {
          claim_id: claimId,
          mode,
          generation: fencingToken.generation,
          expires_at: expiresIso,
        },
      });

      return record;
    });
  }

  /**
   * Send heartbeat to keep an active claim refreshed.
   */
  async heartbeat(claimId: string, agentId: string): Promise<ClaimRecord> {
    return await this.lockManager.withLock('claims_mutex', async () => {
      const claim = await this.getClaim(claimId);
      if (!claim) {
        throw new Error(`Claim not found: ${claimId}`);
      }

      if (claim.owner_agent !== agentId) {
        throw new Error(`Unauthorized heartbeat: Claim ${claimId} is owned by "${claim.owner_agent}", not "${agentId}"`);
      }

      if (claim.status === 'released' || claim.status === 'revoked') {
        throw new Error(`Cannot heartbeat claim in "${claim.status}" status`);
      }

      const nowMs = Date.now();
      const nowIso = new Date(nowMs).toISOString();
      const newExpiresIso = new Date(nowMs + claim.lease_duration_ms).toISOString();

      claim.status = 'active';
      claim.heartbeat_at = nowIso;
      claim.expires_at = newExpiresIso;

      fs.writeFileSync(this.getClaimPath(claimId), JSON.stringify(claim, null, 2), 'utf-8');

      await this.mailboxManager.recordEvent({
        event_id: crypto.randomUUID(),
        timestamp: nowIso,
        event_type: 'claim_heartbeat',
        actor: agentId,
        resource: claim.resource_path,
        details: {
          claim_id: claimId,
          expires_at: newExpiresIso,
        },
      });

      return claim;
    });
  }

  /**
   * Release a held claim.
   */
  async releaseClaim(claimId: string, agentId: string): Promise<boolean> {
    return await this.lockManager.withLock('claims_mutex', async () => {
      const claim = await this.getClaim(claimId);
      if (!claim) {
        return false;
      }

      if (claim.owner_agent !== agentId) {
        throw new Error(`Unauthorized release: Claim ${claimId} is owned by "${claim.owner_agent}", not "${agentId}"`);
      }

      claim.status = 'released';
      fs.writeFileSync(this.getClaimPath(claimId), JSON.stringify(claim, null, 2), 'utf-8');

      await this.mailboxManager.recordEvent({
        event_id: crypto.randomUUID(),
        timestamp: new Date().toISOString(),
        event_type: 'claim_released',
        actor: agentId,
        resource: claim.resource_path,
        details: { claim_id: claimId },
      });

      return true;
    });
  }

  /**
   * Retrieve a claim by ID.
   */
  async getClaim(claimId: string): Promise<ClaimRecord | null> {
    const filePath = this.getClaimPath(claimId);
    if (!fs.existsSync(filePath)) {
      return null;
    }
    try {
      const content = fs.readFileSync(filePath, 'utf-8');
      return JSON.parse(content) as ClaimRecord;
    } catch {
      return null;
    }
  }

  /**
   * Scan and list claims with optional filtering.
   */
  async listClaims(filter: ClaimFilter = {}): Promise<ClaimRecord[]> {
    if (!fs.existsSync(this.baseDir)) {
      return [];
    }

    const files = fs.readdirSync(this.baseDir).filter(f => f.endsWith('.json'));
    const records: ClaimRecord[] = [];
    const nowMs = Date.now();

    for (const file of files) {
      try {
        const content = fs.readFileSync(path.join(this.baseDir, file), 'utf-8');
        const record = JSON.parse(content) as ClaimRecord;

        // Check if active lease has expired in real-time
        if (record.status === 'active' && new Date(record.expires_at).getTime() < nowMs) {
          record.status = 'expired';
        }

        if (filter.owner_agent && record.owner_agent !== filter.owner_agent) continue;
        if (filter.resource_path && record.resource_path !== ClaimManager.normalizeResourcePath(filter.resource_path)) continue;
        if (filter.mode && record.mode !== filter.mode) continue;
        if (filter.status && record.status !== filter.status) continue;
        if (!filter.include_expired && filter.status !== 'expired' && record.status === 'expired') continue;

        records.push(record);
      } catch {
        // Skip corrupted entries
      }
    }

    return records;
  }

  private async listActiveClaimsInternal(): Promise<ClaimRecord[]> {
    const nowMs = Date.now();
    const all = await this.listClaims({ include_expired: true });
    return all.filter(c => c.status === 'active' && new Date(c.expires_at).getTime() >= nowMs);
  }

  /**
   * Auto-Arbiter: Scans and marks expired leases.
   */
  async reapExpiredLeases(): Promise<ClaimRecord[]> {
    return await this.lockManager.withLock('claims_mutex', async () => {
      const files = fs.readdirSync(this.baseDir).filter(f => f.endsWith('.json'));
      const nowMs = Date.now();
      const expired: ClaimRecord[] = [];

      for (const file of files) {
        const filePath = path.join(this.baseDir, file);
        try {
          const content = fs.readFileSync(filePath, 'utf-8');
          const record = JSON.parse(content) as ClaimRecord;

          if (record.status === 'active' && new Date(record.expires_at).getTime() < nowMs) {
            record.status = 'expired';
            fs.writeFileSync(filePath, JSON.stringify(record, null, 2), 'utf-8');
            expired.push(record);

            await this.mailboxManager.recordEvent({
              event_id: crypto.randomUUID(),
              timestamp: new Date().toISOString(),
              event_type: 'claim_expired',
              actor: 'auto_arbiter',
              resource: record.resource_path,
              details: {
                claim_id: record.claim_id,
                expired_owner: record.owner_agent,
                expired_at: record.expires_at,
              },
            });
          }
        } catch {
          // Ignore
        }
      }

      return expired;
    });
  }
}

File: python/nymrel_swarm_protocol/claims.py
"""
nymrel_swarm_protocol - Resource Claims & Auto-Expiring Leases
Copyright (c) 2026 Nymrel / JalenBuilds LLC. Licensed under the MIT License.
"""

import os
import time
import json
import uuid
from typing import Optional, Dict, Any, List, Literal
from dataclasses import dataclass, asdict
from .fencing import AtomicLockManager, FencingClock, iso_now
from .bus import FileMailboxManager, BusEvent

ClaimMode = Literal["exclusive", "shared"]
ClaimStatus = Literal["active", "expired", "released", "revoked"]


@dataclass
class ClaimRecord:
    claim_id: str
    resource_path: str
    owner_agent: str
    mode: ClaimMode
    status: ClaimStatus
    fencing_generation: int
    lease_duration_ms: int
    acquired_at: str
    expires_at: str
    heartbeat_at: str
    metadata: Optional[Dict[str, Any]] = None

    def to_dict(self) -> Dict[str, Any]:
        d = asdict(self)
        if self.metadata is None:
            d["metadata"] = {}
        return d


class ClaimManager:
    def __init__(self, swarm_root: str, default_lease_ms: int = 30000):
        self.base_dir = os.path.join(swarm_root, "claims")
        self.lock_manager = AtomicLockManager(swarm_root)
        self.fencing_clock = FencingClock(swarm_root)
        self.mailbox_manager = FileMailboxManager(swarm_root)
        self.default_lease_ms = default_lease_ms

        os.makedirs(self.base_dir, exist_ok=True)

    def _get_claim_path(self, claim_id: str) -> str:
        return os.path.join(self.base_dir, f"{claim_id}.json")

    @staticmethod
    def normalize_resource_path(resource_path: str) -> str:
        normalized = resource_path.replace("\\", "/").strip()
        if normalized.endswith("/") and len(normalized) > 1:
            normalized = normalized[:-1]
        return normalized.lower()

    @staticmethod
    def is_sub_path(parent: str, child: str) -> bool:
        if parent == child:
            return True
        parent_slash = parent if parent.endswith("/") else f"{parent}/"
        return child.startswith(parent_slash)

    @staticmethod
    def check_conflict(
        path_a: str,
        mode_a: ClaimMode,
        path_b: str,
        mode_b: ClaimMode,
    ) -> bool:
        is_related = ClaimManager.is_sub_path(path_a, path_b) or ClaimManager.is_sub_path(path_b, path_a)
        if not is_related:
            return False

        if mode_a == "exclusive" or mode_b == "exclusive":
            return True

        return False

    def acquire_claim(
        self,
        resource_path: str,
        owner_agent: str,
        mode: ClaimMode = "exclusive",
        lease_duration_ms: Optional[int] = None,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> ClaimRecord:
        def _op():
            norm_path = ClaimManager.normalize_resource_path(resource_path)
            duration = lease_duration_ms if lease_duration_ms is not None else self.default_lease_ms
            now_iso = iso_now()
            expires_iso = iso_now(duration)

            active_claims = self._list_active_internal()

            for existing in active_claims:
                if existing.owner_agent == owner_agent and existing.resource_path == norm_path:
                    # Refresh existing claim
                    existing.lease_duration_ms = duration
                    existing.expires_at = expires_iso
                    existing.heartbeat_at = now_iso
                    if metadata:
                        existing.metadata = {**(existing.metadata or {}), **metadata}
                    with open(self._get_claim_path(existing.claim_id), "w", encoding="utf-8") as f:
                        json.dump(existing.to_dict(), f, indent=2)
                    return existing

                conflict = ClaimManager.check_conflict(
                    existing.resource_path,
                    existing.mode,
                    norm_path,
                    mode,
                )

                if conflict:
                    raise ValueError(
                        f'Claim conflict on "{resource_path}": Already held by "{existing.owner_agent}" '
                        f'({existing.mode} mode, claim_id: {existing.claim_id}, expires: {existing.expires_at})'
                    )

            fencing_token = self.fencing_clock.increment_generation(norm_path, owner_agent)
            claim_id = str(uuid.uuid4())

            record = ClaimRecord(
                claim_id=claim_id,
                resource_path=norm_path,
                owner_agent=owner_agent,
                mode=mode,
                status="active",
                fencing_generation=fencing_token.generation,
                lease_duration_ms=duration,
                acquired_at=now_iso,
                expires_at=expires_iso,
                heartbeat_at=now_iso,
                metadata=metadata or {},
            )

            with open(self._get_claim_path(claim_id), "w", encoding="utf-8") as f:
                json.dump(record.to_dict(), f, indent=2)

            self.mailbox_manager.record_event(BusEvent(
                event_id=str(uuid.uuid4()),
                timestamp=now_iso,
                event_type="claim_acquired",
                actor=owner_agent,
                resource=norm_path,
                details={
                    "claim_id": claim_id,
                    "mode": mode,
                    "generation": fencing_token.generation,
                    "expires_at": expires_iso,
                }
            ))

            return record

        return self.lock_manager.with_lock("claims_mutex", _op)

    def heartbeat(self, claim_id: str, agent_id: str) -> ClaimRecord:
        def _op():
            claim = self.get_claim(claim_id)
            if not claim:
                raise ValueError(f"Claim not found: {claim_id}")

            if claim.owner_agent != agent_id:
                raise PermissionError(f'Unauthorized heartbeat: Claim {claim_id} is owned by "{claim.owner_agent}", not "{agent_id}"')

            if claim.status in ("released", "revoked"):
                raise ValueError(f'Cannot heartbeat claim in "{claim.status}" status')

            now_iso = iso_now()
            new_expires_iso = iso_now(claim.lease_duration_ms)

            claim.status = "active"
            claim.heartbeat_at = now_iso
            claim.expires_at = new_expires_iso

            with open(self._get_claim_path(claim_id), "w", encoding="utf-8") as f:
                json.dump(claim.to_dict(), f, indent=2)

            self.mailbox_manager.record_event(BusEvent(
                event_id=str(uuid.uuid4()),
                timestamp=now_iso,
                event_type="claim_heartbeat",
                actor=agent_id,
                resource=claim.resource_path,
                details={
                    "claim_id": claim_id,
                    "expires_at": new_expires_iso,
                }
            ))

            return claim

        return self.lock_manager.with_lock("claims_mutex", _op)

    def release_claim(self, claim_id: str, agent_id: str) -> bool:
        def _op():
            claim = self.get_claim(claim_id)
            if not claim:
                return False

            if claim.owner_agent != agent_id:
                raise PermissionError(f'Unauthorized release: Claim {claim_id} is owned by "{claim.owner_agent}", not "{agent_id}"')

            claim.status = "released"
            with open(self._get_claim_path(claim_id), "w", encoding="utf-8") as f:
                json.dump(claim.to_dict(), f, indent=2)

            now_iso = iso_now()
            self.mailbox_manager.record_event(BusEvent(
                event_id=str(uuid.uuid4()),
                timestamp=now_iso,
                event_type="claim_released",
                actor=agent_id,
                resource=claim.resource_path,
                details={"claim_id": claim_id}
            ))

            return True

        return self.lock_manager.with_lock("claims_mutex", _op)

    def get_claim(self, claim_id: str) -> Optional[ClaimRecord]:
        file_path = self._get_claim_path(claim_id)
        if not os.path.exists(file_path):
            return None
        try:
            with open(file_path, "r", encoding="utf-8") as f:
                data = json.load(f)
            return ClaimRecord(**data)
        except Exception:
            return None

    def list_claims(
        self,
        owner_agent: Optional[str] = None,
        resource_path: Optional[str] = None,
        mode: Optional[ClaimMode] = None,
        status: Optional[ClaimStatus] = None,
        include_expired: bool = False,
    ) -> List[ClaimRecord]:
        if not os.path.exists(self.base_dir):
            return []

        files = [f for f in os.listdir(self.base_dir) if f.endswith(".json")]
        records: List[ClaimRecord] = []
        now_iso = iso_now()

        for file in files:
            try:
                with open(os.path.join(self.base_dir, file), "r", encoding="utf-8") as f:
                    data = json.load(f)
                record = ClaimRecord(**data)

                if record.status == "active" and record.expires_at < now_iso:
                    record.status = "expired"

                if owner_agent and record.owner_agent != owner_agent:
                    continue
                if resource_path and record.resource_path != self.normalize_resource_path(resource_path):
                    continue
                if mode and record.mode != mode:
                    continue
                if status and record.status != status:
                    continue
                if not include_expired and status != "expired" and record.status == "expired":
                    continue

                records.append(record)
            except Exception:
                pass

        return records

    def _list_active_internal(self) -> List[ClaimRecord]:
        now_iso = iso_now()
        all_claims = self.list_claims(include_expired=True)
        return [c for c in all_claims if c.status == "active" and c.expires_at >= now_iso]

    def reap_expired_leases(self) -> List[ClaimRecord]:
        def _op():
            files = [f for f in os.listdir(self.base_dir) if f.endswith(".json")]
            now_iso = iso_now()
            expired: List[ClaimRecord] = []

            for file in files:
                file_path = os.path.join(self.base_dir, file)
                try:
                    with open(file_path, "r", encoding="utf-8") as f:
                        data = json.load(f)
                    record = ClaimRecord(**data)

                    if record.status == "active" and record.expires_at < now_iso:
                        record.status = "expired"
                        with open(file_path, "w", encoding="utf-8") as f:
                            json.dump(record.to_dict(), f, indent=2)
                        expired.append(record)

                        self.mailbox_manager.record_event(BusEvent(
                            event_id=str(uuid.uuid4()),
                            timestamp=now_iso,
                            event_type="claim_expired",
                            actor="auto_arbiter",
                            resource=record.resource_path,
                            details={
                                "claim_id": record.claim_id,
                                "expired_owner": record.owner_agent,
                                "expired_at": record.expires_at,
                            }
                        ))
                except Exception:
                    pass

            return expired

        return self.lock_manager.with_lock("claims_mutex", _op)
